lib/lints/avoid_unnecessary_type_assertions/avoid_unnecessary_type_assertions_rule.dart

```diff
@@ -0,0 +1,183 @@
+import 'package:custom_lint_builder/custom_lint_builder.dart';
+import 'package:solid_lints/models/rule_config.dart';
+import 'package:solid_lints/models/solid_lint_rule.dart';
+import 'package:solid_lints/utils/types_utils.dart';
+
+part 'avoid_unnecessary_type_assertions_fix.dart';
+
+/// A `avoid-unnecessary-type-assertions` rule which
+/// warns about unnecessary usage of `is` and `whereType` operators
+
+class AvoidUnnecessaryTypeAssertions extends SolidLintRule {
+  /// The [LintCode] of this lint rule that represents
+  /// the error whether we use bad formatted double literals.
+  static const lintName = 'avoid-unnecessary-type-assertions';
+
+  static const _unnecessaryIsCode = LintCode(
+    name: lintName,
+    problemMessage: "Unnecessary usage of the 'is' operator.",
+  );
+
+  static const _unnecessaryWhereTypeCode = LintCode(
+    name: lintName,
+    problemMessage: "Unnecessary usage of the 'whereType'.",
+  );
+
+  AvoidUnnecessaryTypeAssertions._(super.config);
+
+  /// Creates a new instance of [AvoidUnnecessaryTypeAssertions]
+  /// based on the lint configuration.
+  factory AvoidUnnecessaryTypeAssertions.createRule(CustomLintConfigs configs) {
+    final rule = RuleConfig(
+      configs: configs,
+      name: lintName,
+      problemMessage: (_) =>
+          "Unnecessary usage of 'is' or 'whereType' operators.",
+    );
+
+    return AvoidUnnecessaryTypeAssertions._(rule);
+  }
+
+  @override
+  void run(
+    CustomLintResolver resolver,
+    ErrorReporter reporter,
+    CustomLintContext context,
+  ) {
+    context.registry.addIsExpression((node) {
+      if (_isUnnecessaryIsExpression(node)) {
+        reporter.reportErrorForNode(_unnecessaryIsCode, node);
+      }
+    });
+
+    context.registry.addMethodInvocation((node) {
+      if (_isUnnecessaryWhereType(node)) {
+        reporter.reportErrorForNode(_unnecessaryWhereTypeCode, node);
+      }
+    });
+  }
+
+  @override
+  List<Fix> getFixes() => [_UnnecessaryTypeAssertionsFix()];
+
+  bool _isUnnecessaryIsExpression(IsExpression node) {
+    final objectType = node.expression.staticType;
+    final castedType = node.type.type;
+
+    if (node.notOperator != null) {
+      if (objectType != null &&
+          objectType is! TypeParameterType &&
+          objectType is! DynamicType &&
+          !objectType.isDartCoreObject) {
+        return _isUnnecessaryTypeCheck(
+          objectType,
+          castedType,
+          isReversed: true,
+        );
+      }
+    }
+
+    return _isUnnecessaryTypeCheck(objectType, castedType);
+  }
+
+  bool _isUnnecessaryWhereType(MethodInvocation node) {
+    const whereTypeMethodName = 'whereType';
+
+    final targetType = node.target?.staticType;
+
+    if (node.methodName.name == whereTypeMethodName &&
+        isIterableOrSubclass(node.realTarget?.staticType) &&
+        targetType is ParameterizedType) {
+      final isTargetTypeHasGeneric = targetType.typeArguments.isNotEmpty;
```

Comment:
The fact that it's generic is established above. This checks for _actual_ presence of explicit type args.
```suggestion
      final targetHasExplicitTypeArgs = targetType.typeArguments.isNotEmpty;
```